callbacks/opentelemetry.go

```diff
@@ -0,0 +1,288 @@
+	llmCompletionsIndex             = attribute.Key("gen_ai.completion.index")
+	llmCompletionsContent           = attribute.Key("gen_ai.completion.content")
+	llmCompletionsStopReason        = attribute.Key("gen_ai.completion.finish_reasons")
+	llmCompletionsGenerationInfo    = attribute.Key("gen_ai.completion.generation_info")
+	llmCompletionsFuncCallName      = attribute.Key("gen_ai.completion.func_call.name")
+	llmCompletionsFuncCallArguments = attribute.Key("gen_ai.completion.func_call.arguments")
+	llmCompletionsToolCalls         = attribute.Key("gen_ai.completion.tool_calls")
+	llmUsageTotalTokens             = attribute.Key("gen_ai.usage.total_tokens")
+	llmUsagePromptTokens            = attribute.Key("gen_ai.usage.prompt_tokens")
+	llmUsageCompletionTokens        = attribute.Key("gen_ai.usage.completion_tokens")
+)
+
+// semconvGenAIMessage is a struct that represents a message to/from a Gen-AI model in accordance with the semantic conventions for Gen-AI LLM spans.
+type semconvGenAIMessage struct {
+	Index   int
+	Role    string
+	Content string
+}
+
+// OpenTelemetryCallbacksHandlerOptions contains the options for the OpenTelemetryCallbacksHandler.
+type OpenTelemetryCallbacksHandlerOptions struct {
+	logPrompts     bool
+	logCompletions bool
+}
+
+// OpenTelemetryCallbacksHandlerOption is a function that can be used to modify the behavior of the OpenTelemetryCallbacksHandler. An arbitrary number of such functions can be passed to NewOpenTelemetryCallbacksHandler.
+type OpenTelemetryCallbacksHandlerOption func(*OpenTelemetryCallbacksHandlerOptions) *OpenTelemetryCallbacksHandlerOptions
+
+// WithLogPrompts is an option to enable logging of prompts in the OpenTelemetryCallbacksHandler.
+func WithLogPrompts(logPrompts bool) OpenTelemetryCallbacksHandlerOption {
+	return func(opts *OpenTelemetryCallbacksHandlerOptions) *OpenTelemetryCallbacksHandlerOptions {
+		opts.logPrompts = logPrompts
+		return opts
+	}
+}
+
+// WithLogCompletions is an option to enable logging of completions in the OpenTelemetryCallbacksHandler.
+func WithLogCompletions(logCompletions bool) OpenTelemetryCallbacksHandlerOption {
+	return func(opts *OpenTelemetryCallbacksHandlerOptions) *OpenTelemetryCallbacksHandlerOptions {
+		opts.logCompletions = logCompletions
+		return opts
+	}
+}
+
+var _ Handler = &OpenTelemetryCallbacksHandler{}
+
+// HandleAgentAction implements Handler.
+func (o *OpenTelemetryCallbacksHandler) HandleAgentAction(_ context.Context, _ schema.AgentAction) {
+}
+
+// HandleAgentFinish implements Handler.
+func (o *OpenTelemetryCallbacksHandler) HandleAgentFinish(_ context.Context, _ schema.AgentFinish) {
+}
+
+// HandleChain implements Handler.
+func (o *OpenTelemetryCallbacksHandler) HandleChain(ctx context.Context, _ map[string]any, info schema.ChainInfo, next func(ctx context.Context) (map[string]any, error)) (map[string]any, error) {
+	ctx, span := o.tracer.Start(ctx, "langchaingo.chain."+info.Name)
+	defer span.End()
+	return next(ctx)
```

Comment:
If there is an error here, it could be nice to add it to the span?
https://opentelemetry.io/docs/languages/go/instrumentation/#record-errors

Reply:
[  ] Will do